src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs

```diff
@@ -206,41 +210,53 @@ public async Task StopAsync(bool processRemainingBlocks = false)
             }
 
             await Task.WhenAll((_recoveryTask ?? Task.CompletedTask), (_processorTask ?? Task.CompletedTask));
-            if (_logger.IsInfo)
-                _logger.Info("Blockchain Processor shutdown complete.. please wait for all components to close");
+            if (_logger.IsInfo) _logger.Info("Blockchain Processor shutdown complete.. please wait for all components to close");
         }
 
         private void RunRecoveryLoop()
         {
-            if (_logger.IsDebug)
-                _logger.Debug($"Starting recovery loop - {_blockQueue.Count} blocks waiting in the queue.");
+            if (_logger.IsDebug) _logger.Debug($"Starting recovery loop - {_blockQueue.Count} blocks waiting in the queue.");
             _lastProcessedBlock = DateTime.UtcNow;
             foreach (BlockRef blockRef in _recoveryQueue.GetConsumingEnumerable(_loopCancellationSource.Token))
             {
-                if (blockRef.Resolve(_blockTree))
+                try
                 {
-                    Interlocked.Add(ref _currentRecoveryQueueSize, -blockRef.Block!.Transactions.Length);
-                    if (_logger.IsTrace)
-                        _logger.Trace(
-                            $"Recovering addresses for block {blockRef.BlockHash?.ToString() ?? blockRef.Block.ToString(Block.Format.Short)}.");
-                    _recoveryStep.RecoverData(blockRef.Block);
-
-                    try
+                    if (blockRef.Resolve(_blockTree))
                     {
-                        _blockQueue.Add(blockRef);
+                        Interlocked.Add(ref _currentRecoveryQueueSize, -blockRef.Block!.Transactions.Length);
+                        if (_logger.IsTrace) _logger.Trace($"Recovering addresses for block {blockRef.BlockHash}.");
+                        _recoveryStep.RecoverData(blockRef.Block);
+
+                        try
+                        {
+                            _blockQueue.Add(blockRef);
+                        }
+                        catch (Exception e)
+                        {
+                            Interlocked.Decrement(ref _queueCount);
```

Comment:
Shouldnt you put this in a finally block?

Reply:
No, if adding is successful the counter isn't decremented.